examples/multi_gpu/distributed_batching.py

```diff
@@ -84,18 +84,19 @@ def run(rank, world_size: int, dataset_name: str, root: str):
     for epoch in range(1, 51):
         model.train()
 
-        total_loss = 0
+        total_loss = torch.zeros(2).to(rank)
         for data in train_loader:
             data = data.to(rank)
             optimizer.zero_grad()
             logits = model(data.x, data.adj_t, data.batch)
             loss = criterion(logits, data.y.to(torch.float))
             loss.backward()
             optimizer.step()
-            total_loss += float(loss) * logits.size(0)
-        loss = total_loss / len(train_loader.dataset)
+            total_loss[0] += loss.item() * logits.size(0)
+            total_loss[1] += len(data)
 
-        dist.barrier()
+        dist.all_reduce(total_loss, op=dist.ReduceOp.SUM)
```

Comment:
Does `all_reduce` remove the need for calling `barrier`?

Reply:
I assumed that this is the case because it is all to all communication and 

from https://pytorch.org/docs/stable/_modules/torch/distributed/distributed_c10d.html#all_reduce
```
def all_reduce(tensor, op=ReduceOp.SUM, group=None, **async_op=False**):
    """
    Reduces the tensor data across all machines in such a way that all get
    the final result.

    After the call ``tensor`` is going to be bitwise identical in all processes.

```

but after the all_reduce there is a block where only rank 0 does work and then there is another barrier in the end where the ranks should be synchronised again.